File: python/features/steps/standup_panel_steps.py
"""Behave steps for console standup drawer scenarios."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path

from behave import given, then, when

from kanbus.models import IssueData
from kanbus.standup import (
    MEETING_SCRIPT_PROFILE,
    build_standup_report,
    collect_right_now_texts,
    format_standup_text,
    resolve_standup_profile,
)
from kanbus.standup_window import (
    DEFAULT_STANDUP_LOOKBACK,
    ROLLING_WINDOW,
    StandupWindowSettings,
)
from zoneinfo import ZoneInfo
from kanbus.standup_command import STANDUP_DEFAULT_STATUS_FILTER

from features.steps.console_ui_steps import (
    ConsoleIssue,
    ConsoleState,
    _console_app_root,
    _require_console_state,
)


@dataclass
class StandupPanelState:
    """Simulated standup drawer state for console UI scenarios."""

    is_open: bool = False
    profile: str = MEETING_SCRIPT_PROFILE
    is_generating: bool = False
    error: str | None = None
    report_text: str | None = None
    section_names: list[str] = field(default_factory=list)
    clipboard: str | None = None
    generation_should_fail: bool = False


def _ensure_standup_state(context: object) -> StandupPanelState:
    state = getattr(context, "console_standup_state", None)
    if state is None:
        state = StandupPanelState()
        context.console_standup_state = state
    return state


def _parse_console_timestamp(value: str | None) -> datetime:
    if not value:
        return datetime.now(timezone.utc)
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def _standup_statuses() -> set[str]:
    return {status.strip() for status in STANDUP_DEFAULT_STATUS_FILTER.split(",")}


def _console_issue_to_issue_data(issue: ConsoleIssue) -> IssueData:
    timestamp = _parse_console_timestamp(issue.updated_at)
    return IssueData.model_validate(
        {
            "id": issue.identifier or issue.title,
            "title": issue.title,
            "description": "",
            "type": issue.issue_type,
            "status": issue.status,
            "priority": issue.priority,
            "assignee": issue.assignee,
            "creator": "agent",
            "parent": None,
            "labels": [],
            "dependencies": [],
            "comments": [],
            "created_at": timestamp,
            "updated_at": timestamp,
            "closed_at": (
                _parse_console_timestamp(issue.closed_at) if issue.closed_at else None
            ),
            "right_now_summary": issue.right_now_summary,
            "right_now_updated_at": timestamp if issue.right_now_summary else None,
            "custom": {},
            "agent": None,
        }
    )


def _standup_fact_feed_issues(console_state: ConsoleState) -> list[IssueData]:
    allowed_statuses = _standup_statuses()
    issues = [
        _console_issue_to_issue_data(issue)
        for issue in console_state.issues
        if issue.status in allowed_statuses
    ]
    return sorted(
        issues,
        key=lambda issue: (
            issue.updated_at.isoformat() if issue.updated_at else "",
            issue.identifier,
        ),
        reverse=True,
    )


def _generate_from_console_state(
    console_state: ConsoleState,
    profile_name: str,
) -> tuple[str, list[str]]:
    profile = resolve_standup_profile(profile_name)
    issues = _standup_fact_feed_issues(console_state)
    right_now_texts = collect_right_now_texts(issues)
    window_settings = StandupWindowSettings(
        window=ROLLING_WINDOW,
        lookback=DEFAULT_STANDUP_LOOKBACK,
        lookback_hours=24,
        skip_weekends=False,
        timezone=ZoneInfo("UTC"),
    )
    from kanbus.standup import load_standup_configuration
    from kanbus.standup_rollup import resolve_standup_rollup

    configuration = load_standup_configuration(Path(context.working_directory))
    rollup_settings = resolve_standup_rollup(None, configuration, False)
    report = build_standup_report(
        Path(context.working_directory),
        profile,
        issues,
        right_now_texts,
        {},
        datetime.now(timezone.utc),
        window_settings,
        configuration,
        rollup_settings,
        False,
    )
    section_names = [section.name for section in report.sections]
    return format_standup_text(report), section_names


@given("standup generation is configured to fail")
def given_standup_generation_configured_to_fail(context: object) -> None:
    """Configure the simulated standup drawer to fail generation.

    :param context: Behave context object.
    :type context: object
    """
    _ensure_standup_state(context).generation_should_fail = True


@when("I open the standup drawer")
def when_open_standup_drawer(context: object) -> None:
    """Open the standup drawer from the Now panel.

    :param context: Behave context object.
    :type context: object
    """
    _ensure_standup_state(context).is_open = True


@when('I select the standup profile "{profile}"')
def when_select_standup_profile(context: object, profile: str) -> None:
    """Select a standup profile in the drawer.

    :param context: Behave context object.
    :type context: object
    :param profile: Standup profile identifier.
    :type profile: str
    """
    _ensure_standup_state(context).profile = profile


@when("I generate the standup report")
def when_generate_standup_report(context: object) -> None:
    """Generate a standup report in the simulated drawer.

    :param context: Behave context object.
    :type context: object
    """
    standup = _ensure_standup_state(context)
    console_state = _require_console_state(context)
    standup.is_generating = True
    standup.error = None
    standup.report_text = None
    standup.section_names = []
    if standup.generation_should_fail:
        standup.is_generating = False
        standup.error = "standup generation failed"
        return
    try:
        report_text, section_names = _generate_from_console_state(
            console_state,
            standup.profile,
        )
    except Exception as error:
        standup.is_generating = False
        standup.error = str(error)
        return
    standup.report_text = report_text
    standup.section_names = section_names
    standup.is_generating = False


@when("I copy the standup report")
def when_copy_standup_report(context: object) -> None:
    """Copy the generated standup report to the simulated clipboard.

    :param context: Behave context object.
    :type context: object
    """
    standup = _ensure_standup_state(context)
    if standup.report_text is None:
        raise AssertionError("no standup report to copy")
    standup.clipboard = standup.report_text


@when('I request a standup report from the console API with profile "{profile}"')
def when_request_standup_from_console_api(context: object, profile: str) -> None:
    """POST /api/standup against the running console server.

    :param context: Behave context object.
    :type context: object
    :param profile: Standup profile identifier.
    :type profile: str
    """
    port = getattr(context, "console_server_port", None)
    if port is None:
        raise AssertionError("console server is not running")
    url = f"http://127.0.0.1:{port}/api/standup"
    payload = json.dumps({"profile": profile}).encode()
    request = urllib.request.Request(
        url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=30) as response:
            body = response.read().decode()
    except urllib.error.HTTPError as error:
        body = error.read().decode()
        context.standup_api_status = error.code
        context.standup_api_response = json.loads(body) if body else {}
        return
    context.standup_api_status = 200
    context.standup_api_response = json.loads(body)


def _standup_drawer_source() -> str:
    return (
        _console_app_root() / "src" / "components" / "StandupDrawer.tsx"
    ).read_text()


def _assert_standup_drawer_portals_to_body() -> None:
    source = _standup_drawer_source()
    if "createPortal" not in source or "document.body" not in source:
        raise AssertionError(
            "StandupDrawer must render via createPortal to document.body"
        )


@given("the browser viewport is {width:d} by {height:d}")
def given_browser_viewport(context: object, width: int, height: int) -> None:
    """Record a viewport size for simulated console UI scenarios.

    :param context: Behave context object.
    :type context: object
    :param width: Viewport width in CSS pixels.
    :type width: int
    :param height: Viewport height in CSS pixels.
    :type height: int
    """
    context.console_viewport = {"width": width, "height": height}


@then("the standup drawer should be in the viewport")
def then_standup_drawer_in_viewport(context: object) -> None:
    """Verify the standup drawer is viewport-positioned, not view-track-offset.

    :param context: Behave context object.
    :type context: object
    """
    standup = _ensure_standup_state(context)
    if not standup.is_open:
        raise AssertionError("expected standup drawer to be open")
    _assert_standup_drawer_portals_to_body()


@then("the standup profile select should be in the viewport")
def then_standup_profile_select_in_viewport(context: object) -> None:
    """Verify standup profile control is reachable in the viewport layout.

    :param context: Behave context object.
    :type context: object
    """
    _assert_standup_drawer_portals_to_body()


@then("the standup window select should be in the viewport")
def then_standup_window_select_in_viewport(context: object) -> None:
    """Verify standup window control is reachable in the viewport layout.

    :param context: Behave context object.
    :type context: object
    """
    _assert_standup_drawer_portals_to_body()


@then("the standup lookback input should be in the viewport")
def then_standup_lookback_input_in_viewport(context: object) -> None:
    """Verify standup lookback control is reachable in the viewport layout.

    :param context: Behave context object.
    :type context: object
    """
    _assert_standup_drawer_portals_to_body()


@then("the standup skip weekends checkbox should be in the viewport")
def then_standup_skip_weekends_checkbox_in_viewport(context: object) -> None:
    """Verify standup skip-weekends control is reachable in the viewport layout.

    :param context: Behave context object.
    :type context: object
    """
    _assert_standup_drawer_portals_to_body()


@then("the now standup button should be visible")
def then_now_standup_button_visible(context: object) -> None:
    """Verify the Now panel toolbar exposes the Standup button.

    :param context: Behave context object.
    :type context: object
    """
    state = _require_console_state(context)
    if state.panel_mode != "now":
        raise AssertionError("expected Now panel to be active")
    source = (
        _console_app_root() / "src" / "components" / "CurrentStatusPanel.tsx"
    ).read_text()
    if 'data-testid="now-standup-button"' not in source:
        raise AssertionError("CurrentStatusPanel is missing now-standup-button")


@then('the standup drawer should show section "{section_name}"')
def then_standup_drawer_shows_section(context: object, section_name: str) -> None:
    """Verify a standup section heading is visible in the drawer.

    :param context: Behave context object.
    :type context: object
    :param section_name: Expected section heading.
    :type section_name: str
    """
    standup = _ensure_standup_state(context)
    if section_name not in standup.section_names:
        raise AssertionError(
            f"expected section {section_name}, got {standup.section_names}"
        )


@then('the standup drawer result should mention "{text}"')
def then_standup_drawer_result_mentions(context: object, text: str) -> None:
    """Verify generated standup text mentions expected content.

    :param context: Behave context object.
    :type context: object
    :param text: Expected substring.
    :type text: str
    """
    standup = _ensure_standup_state(context)
    if standup.report_text is None or text not in standup.report_text:
        raise AssertionError(f"expected standup result to mention {text!r}")


@then('the standup drawer should show error "{message}"')
def then_standup_drawer_shows_error(context: object, message: str) -> None:
    """Verify the standup drawer surfaces a fail-closed error.

    :param context: Behave context object.
    :type context: object
    :param message: Expected error message.
    :type message: str
    """
    standup = _ensure_standup_state(context)
    if standup.error != message:
        raise AssertionError(f"expected error {message!r}, got {standup.error!r}")


@then("the standup drawer should not show a success result")
def then_standup_drawer_has_no_success_result(context: object) -> None:
    """Verify no successful standup result is shown after failure.

    :param context: Behave context object.
    :type context: object
    """
    standup = _ensure_standup_state(context)
    if standup.report_text:
        raise AssertionError("expected no standup success result")


@then('the standup clipboard should contain "{text}"')
def then_standup_clipboard_contains(context: object, text: str) -> None:
    """Verify copied standup text contains expected content.

    :param context: Behave context object.
    :type context: object
    :param text: Expected substring.
    :type text: str
    """
    standup = _ensure_standup_state(context)
    if standup.clipboard is None or text not in standup.clipboard:
        raise AssertionError(f"expected clipboard to contain {text!r}")


@then('the standup API response should have profile "{profile}"')
def then_standup_api_response_profile(context: object, profile: str) -> None:
    """Verify console API standup response profile.

    :param context: Behave context object.
    :type context: object
    :param profile: Expected profile identifier.
    :type profile: str
    """
    status = getattr(context, "standup_api_status", None)
    if status != 200:
        response = getattr(context, "standup_api_response", {})
        raise AssertionError(f"expected HTTP 200, got {status}: {response!r}")
    response = getattr(context, "standup_api_response", {})
    if response.get("profile") != profile:
        raise AssertionError(
            f"expected profile {profile!r}, got {response.get('profile')!r}"
        )


@then('the standup API response should include section "{section_name}"')
def then_standup_api_response_includes_section(
    context: object, section_name: str
) -> None:
    """Verify console API standup response includes a section heading.

    :param context: Behave context object.
    :type context: object
    :param section_name: Expected section heading.
    :type section_name: str
    """
    response = getattr(context, "standup_api_response", {})
    sections = response.get("sections", [])
    names = [section.get("name") for section in sections]
    if section_name not in names:
        raise AssertionError(f"expected section {section_name!r}, got {names}")


@then('the standup API response text should mention "{text}"')
def then_standup_api_response_text_mentions(context: object, text: str) -> None:
    """Verify console API standup response text mentions expected content.

    :param context: Behave context object.
    :type context: object
    :param text: Expected substring.
    :type text: str
    """
    response = getattr(context, "standup_api_response", {})
    report_text = response.get("text", "")
    if text not in report_text:
        raise AssertionError(f"expected API text to mention {text!r}")

File: rust/features/steps/standup_panel_steps.rs
use std::collections::{BTreeMap, HashMap};
use std::fs;
use std::path::PathBuf;
use std::thread;

use chrono::{DateTime, Utc};
use cucumber::{given, then, when};
use reqwest::blocking::Client;
use serde_json::json;

use chrono_tz::UTC;
use kanbus::models::IssueData;
use kanbus::standup::{
    build_standup_report, collect_right_now_texts, format_standup_text, load_standup_configuration,
    resolve_standup_profile, MEETING_SCRIPT_PROFILE,
};
use kanbus::standup_command::STANDUP_DEFAULT_STATUS_FILTER;
use kanbus::standup_rollup::resolve_standup_rollup;
use kanbus::standup_window::{StandupWindowSettings, DEFAULT_STANDUP_LOOKBACK, ROLLING_WINDOW};

use crate::step_definitions::console_ui_steps::{ConsoleIssue, ConsoleState};
use crate::step_definitions::initialization_steps::KanbusWorld;

#[derive(Debug, Clone, Default)]
pub struct StandupPanelState {
    pub is_open: bool,
    pub profile: String,
    pub is_generating: bool,
    pub error: Option<String>,
    pub report_text: Option<String>,
    pub section_names: Vec<String>,
    pub clipboard: Option<String>,
    pub generation_should_fail: bool,
}

fn ensure_standup_state(world: &mut KanbusWorld) -> &mut StandupPanelState {
    if world.console_standup_state.is_none() {
        world.console_standup_state = Some(StandupPanelState {
            profile: MEETING_SCRIPT_PROFILE.to_string(),
            ..StandupPanelState::default()
        });
    }
    world.console_standup_state.as_mut().unwrap()
}

fn require_console_state(world: &mut KanbusWorld) -> &mut ConsoleState {
    world
        .console_state
        .as_mut()
        .expect("console state not initialized")
}

fn parse_console_timestamp(value: Option<&str>) -> DateTime<Utc> {
    match value {
        Some(raw) => DateTime::parse_from_rfc3339(raw)
            .map(|parsed| parsed.with_timezone(&Utc))
            .unwrap_or_else(|_| Utc::now()),
        None => Utc::now(),
    }
}

fn standup_statuses() -> Vec<String> {
    STANDUP_DEFAULT_STATUS_FILTER
        .split(',')
        .map(|status| status.trim().to_string())
        .collect()
}

fn console_issue_to_issue_data(issue: &ConsoleIssue) -> IssueData {
    let timestamp = parse_console_timestamp(issue.updated_at.as_deref());
    IssueData {
        identifier: issue
            .identifier
            .clone()
            .unwrap_or_else(|| issue.title.clone()),
        title: issue.title.clone(),
        description: String::new(),
        issue_type: issue.issue_type.clone(),
        status: issue.status.clone(),
        priority: issue.priority,
        assignee: issue.assignee.clone(),
        creator: Some("agent".to_string()),
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: issue
            .closed_at
            .as_deref()
            .map(|value| parse_console_timestamp(Some(value))),
        right_now_summary: issue.right_now_summary.clone(),
        right_now_updated_at: issue.right_now_summary.as_ref().map(|_| timestamp),
        custom: BTreeMap::new(),
        agent: None,
    }
}

fn standup_fact_feed_issues(console_state: &ConsoleState) -> Vec<IssueData> {
    let allowed = standup_statuses();
    let mut issues = console_state
        .issues
        .iter()
        .filter(|issue| allowed.iter().any(|status| status == &issue.status))
        .map(console_issue_to_issue_data)
        .collect::<Vec<_>>();
    issues.sort_by(|left, right| {
        right
            .updated_at
            .cmp(&left.updated_at)
            .then_with(|| left.identifier.cmp(&right.identifier))
    });
    issues
}

fn generate_from_console_state(
    console_state: &ConsoleState,
    profile_name: &str,
    root: &std::path::Path,
) -> Result<(String, Vec<String>), String> {
    let profile = resolve_standup_profile(Some(profile_name)).map_err(|error| error.to_string())?;
    let issues = standup_fact_feed_issues(console_state);
    let right_now_texts = collect_right_now_texts(&issues).map_err(|error| error.to_string())?;
    let window_settings = StandupWindowSettings {
        window: ROLLING_WINDOW.to_string(),
        lookback: DEFAULT_STANDUP_LOOKBACK.to_string(),
        lookback_hours: 24,
        skip_weekends: false,
        timezone: UTC,
    };
    let configuration = load_standup_configuration(root).map_err(|error| error.to_string())?;
    let rollup_settings =
        resolve_standup_rollup(None, &configuration, false).map_err(|error| error.to_string())?;
    let report = build_standup_report(
        root,
        &profile,
        &issues,
        &right_now_texts,
        &HashMap::new(),
        Utc::now(),
        &window_settings,
        false,
        &configuration,
        &rollup_settings,
    )
    .map_err(|error| error.to_string())?;
    let section_names = report
        .sections
        .iter()
        .map(|section| section.name.clone())
        .collect();
    Ok((format_standup_text(&report), section_names))
}

fn console_app_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("apps")
        .join("console")
}

fn assert_standup_drawer_portals_to_body() {
    let source = fs::read_to_string(console_app_root().join("src/components/StandupDrawer.tsx"))
        .expect("read StandupDrawer.tsx");
    assert!(
        source.contains("createPortal") && source.contains("document.body"),
        "StandupDrawer must render via createPortal to document.body"
    );
}

#[given(regex = r"^the browser viewport is (?P<width>\d+) by (?P<height>\d+)$")]
fn given_browser_viewport(world: &mut KanbusWorld, width: u32, height: u32) {
    world.console_viewport = Some((width, height));
}

#[given("standup generation is configured to fail")]
fn given_standup_generation_configured_to_fail(world: &mut KanbusWorld) {
    ensure_standup_state(world).generation_should_fail = true;
}

#[when("I open the standup drawer")]
fn when_open_standup_drawer(world: &mut KanbusWorld) {
    ensure_standup_state(world).is_open = true;
}

#[when(regex = r#"^I select the standup profile "(?P<profile>[^"]+)"$"#)]
fn when_select_standup_profile(world: &mut KanbusWorld, profile: String) {
    ensure_standup_state(world).profile = profile;
}

#[when("I generate the standup report")]
fn when_generate_standup_report(world: &mut KanbusWorld) {
    let generation_should_fail = ensure_standup_state(world).generation_should_fail;
    let profile = ensure_standup_state(world).profile.clone();
    let generated = if generation_should_fail {
        Err("standup generation failed".to_string())
    } else {
        let root = world
            .working_directory
            .clone()
            .expect("working directory not initialized");
        let console_state = require_console_state(world);
        generate_from_console_state(console_state, &profile, &root)
    };
    let standup = ensure_standup_state(world);
    standup.is_generating = true;
    standup.error = None;
    standup.report_text = None;
    standup.section_names.clear();
    match generated {
        Ok((report_text, section_names)) => {
            standup.report_text = Some(report_text);
            standup.section_names = section_names;
            standup.is_generating = false;
        }
        Err(message) => {
            standup.error = Some(message);
            standup.is_generating = false;
        }
    }
}

#[when("I copy the standup report")]
fn when_copy_standup_report(world: &mut KanbusWorld) {
    let standup = ensure_standup_state(world);
    let report_text = standup
        .report_text
        .clone()
        .expect("no standup report to copy");
    standup.clipboard = Some(report_text);
}

#[when(
    regex = r#"^I request a standup report from the console API with profile "(?P<profile>[^"]+)"$"#
)]
fn when_request_standup_from_console_api(world: &mut KanbusWorld, profile: String) {
    let port = world.console_port.expect("console server is not running");
    let url = format!("http://127.0.0.1:{port}/api/standup");
    let payload = json!({ "profile": profile });
    let (status, body) = thread::spawn(move || {
        let client = Client::new();
        let response = client
            .post(url)
            .json(&payload)
            .send()
            .expect("standup request");
        let status = response.status().as_u16();
        let body = response.text().expect("standup response body");
        (status, body)
    })
    .join()
    .expect("standup request thread");
    world.standup_api_status = Some(status);
    world.standup_api_response = Some(serde_json::from_str(&body).expect("parse standup json"));
}

#[then("the standup drawer should be in the viewport")]
fn then_standup_drawer_in_viewport(world: &mut KanbusWorld) {
    let standup = ensure_standup_state(world);
    assert!(standup.is_open, "expected standup drawer to be open");
    assert_standup_drawer_portals_to_body();
}

#[then("the standup profile select should be in the viewport")]
fn then_standup_profile_select_in_viewport(_world: &mut KanbusWorld) {
    assert_standup_drawer_portals_to_body();
}

#[then("the standup window select should be in the viewport")]
fn then_standup_window_select_in_viewport(_world: &mut KanbusWorld) {
    assert_standup_drawer_portals_to_body();
}

#[then("the standup lookback input should be in the viewport")]
fn then_standup_lookback_input_in_viewport(_world: &mut KanbusWorld) {
    assert_standup_drawer_portals_to_body();
}

#[then("the standup skip weekends checkbox should be in the viewport")]
fn then_standup_skip_weekends_checkbox_in_viewport(_world: &mut KanbusWorld) {
    assert_standup_drawer_portals_to_body();
}

#[then("the now standup button should be visible")]
fn then_now_standup_button_visible(world: &mut KanbusWorld) {
    let console_state = require_console_state(world);
    assert_eq!(
        console_state.panel_mode, "now",
        "expected Now panel to be active"
    );
    let source =
        fs::read_to_string(console_app_root().join("src/components/CurrentStatusPanel.tsx"))
            .expect("read CurrentStatusPanel.tsx");
    assert!(
        source.contains("data-testid=\"now-standup-button\""),
        "CurrentStatusPanel is missing now-standup-button"
    );
}

#[then(regex = r#"^the standup drawer should show section "(?P<section_name>[^"]+)"$"#)]
fn then_standup_drawer_shows_section(world: &mut KanbusWorld, section_name: String) {
    let standup = ensure_standup_state(world);
    assert!(
        standup
            .section_names
            .iter()
            .any(|name| name == &section_name),
        "expected section {}, got {:?}",
        section_name,
        standup.section_names
    );
}

#[then(regex = r#"^the standup drawer result should mention "(?P<text>[^"]+)"$"#)]
fn then_standup_drawer_result_mentions(world: &mut KanbusWorld, text: String) {
    let standup = ensure_standup_state(world);
    let report_text = standup.report_text.as_deref().unwrap_or("");
    assert!(
        report_text.contains(&text),
        "expected standup result to mention {}",
        text
    );
}

#[then(regex = r#"^the standup drawer should show error "(?P<message>[^"]+)"$"#)]
fn then_standup_drawer_shows_error(world: &mut KanbusWorld, message: String) {
    let standup = ensure_standup_state(world);
    assert_eq!(standup.error.as_deref(), Some(message.as_str()));
}

#[then("the standup drawer should not show a success result")]
fn then_standup_drawer_has_no_success_result(world: &mut KanbusWorld) {
    let standup = ensure_standup_state(world);
    assert!(
        standup.report_text.is_none(),
        "expected no standup success result"
    );
}

#[then(regex = r#"^the standup clipboard should contain "(?P<text>[^"]+)"$"#)]
fn then_standup_clipboard_contains(world: &mut KanbusWorld, text: String) {
    let standup = ensure_standup_state(world);
    let clipboard = standup.clipboard.as_deref().unwrap_or("");
    assert!(
        clipboard.contains(&text),
        "expected clipboard to contain {}",
        text
    );
}

#[then(regex = r#"^the standup API response should have profile "(?P<profile>[^"]+)"$"#)]
fn then_standup_api_response_profile(world: &mut KanbusWorld, profile: String) {
    assert_eq!(world.standup_api_status, Some(200));
    let response = world
        .standup_api_response
        .as_ref()
        .expect("standup api response");
    assert_eq!(
        response.get("profile").and_then(|value| value.as_str()),
        Some(profile.as_str())
    );
}

#[then(regex = r#"^the standup API response should include section "(?P<section_name>[^"]+)"$"#)]
fn then_standup_api_response_includes_section(world: &mut KanbusWorld, section_name: String) {
    let response = world
        .standup_api_response
        .as_ref()
        .expect("standup api response");
    let sections = response
        .get("sections")
        .and_then(|value| value.as_array())
        .expect("sections array");
    let names = sections
        .iter()
        .filter_map(|section| section.get("name").and_then(|name| name.as_str()))
        .collect::<Vec<_>>();
    assert!(
        names.iter().any(|name| *name == section_name),
        "expected section {}, got {:?}",
        section_name,
        names
    );
}

#[then(regex = r#"^the standup API response text should mention "(?P<text>[^"]+)"$"#)]
fn then_standup_api_response_text_mentions(world: &mut KanbusWorld, text: String) {
    let response = world
        .standup_api_response
        .as_ref()
        .expect("standup api response");
    let report_text = response
        .get("text")
        .and_then(|value| value.as_str())
        .unwrap_or("");
    assert!(
        report_text.contains(&text),
        "expected API text to mention {}",
        text
    );
}
